c/auth.c

```diff
@@ -363,6 +369,137 @@ int validate_signature_cardano(void *prefilled_data, const uint8_t *sig,
+
+    uint8_t len_buf[(sizeof(size_t) * 8 + 6) / 7];
+    size_t written_bytes = write_varint((uint8_t *)len_buf, msg_len);
+    keccak_update(&ctx, (uint8_t *)len_buf, written_bytes);
+
+    keccak_update(&ctx, (uint8_t *)msg, msg_len);
+
+    keccak_final(&ctx, (uint8_t *)hash);
+}
+
+void monero_hash_to_scalar(uint8_t *msg, size_t msg_len, uint8_t *key,
+                           uint8_t *comm, uint8_t scalar[32]) {
+    uint8_t state[200];
+    SHA3_CTX sha3_ctx;
+
+    keccak_init(&sha3_ctx);
+    keccak_update(&sha3_ctx, msg, msg_len);
+    keccak_update(&sha3_ctx, key, 32);
+    keccak_update(&sha3_ctx, comm, 32);
+    keccak_final(&sha3_ctx, state);
+    memcpy(scalar, &state, 32);
+    sc_reduce32(scalar);
+}
+
+// See
+// https://github.com/monero-project/monero/blob/e06129bb4d1076f4f2cebabddcee09f1e9e30dcc/src/crypto/crypto.cpp#L319-L341
+int ed25519_verify_monero(const unsigned char *signature,
+                          const unsigned char *message, size_t message_len,
+                          const unsigned char *public_key) {
+    ge_p2 tmp2;
+    ge_p3 tmp3;
+    uint8_t c[32];
+    uint8_t comm[32];
+    uint8_t *sig_c = (uint8_t *)signature;
+    uint8_t *sig_r = sig_c + 32;
+    uint8_t zero[32];
+    uint8_t sig_c_neg[32];
+
+    if (sc_check(sig_c) != 0 || sc_check(sig_r) != 0 || !sc_isnonzero(sig_c)) {
+        return 0;
+    }
+    // TODO: implement ge_frombytes_vartime instead of using
+    // ge_frombytes_negate_vartime and then multiple the result with a negative
+    // scalar
+    sc_0(zero);
+    sc_sub(sig_c_neg, zero, sig_c);
+    if (ge_frombytes_negate_vartime(&tmp3, public_key) != 0) {
+        return 0;
+    }
+    ge_double_scalarmult_vartime(&tmp2, sig_c_neg, &tmp3, sig_r);
+    ge_tobytes(comm, &tmp2);
+
+    static const uint8_t infinity[32] = {1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                                         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                                         0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+    if (memcmp(&comm, &infinity, 32) == 0) return 0;
+    monero_hash_to_scalar((uint8_t *)message, message_len,
+                          (uint8_t *)public_key, comm, c);
+    sc_sub(c, c, sig_c);
+    return sc_isnonzero((const uint8_t *)c) == 0;
+}
+
+int validate_signature_monero(void *prefilled_data, const uint8_t *sig,
+                              size_t sig_len, const uint8_t *msg,
+                              size_t msg_len, uint8_t *output,
+                              size_t *output_len) {
+    int err = 0;
+
+    CHECK2(msg_len == BLAKE2B_BLOCK_SIZE, ERROR_INVALID_ARG);
+    CHECK2(sig_len == MONERO_DATA_SIZE, ERROR_INVALID_ARG);
+
+    uint8_t *mode_ptr = (uint8_t *)sig + MONERO_SIGNATURE_SIZE;
+    CHECK2(*mode_ptr != 0 || *mode_ptr != 1, ERROR_INVALID_ARG);
+
+    uint8_t *spend_pubkey = mode_ptr + sizeof(*mode_ptr);
+    uint8_t *view_pubkey = spend_pubkey + MONERO_PUBKEY_SIZE;
+    uint8_t *pubkey = *mode_ptr == 0 ? spend_pubkey : view_pubkey;
```

Comment:
Why does view key can validate the signature?

Reply:
The official `monero-wallet-cli` supports both signing with spend key and view key.

`monero-wallet-cli --wallet-file ckb-auth-test-wallet --password pw --offline help sign` outputs

```
 help sign
Command usage:
  sign [<account_index>,<address_index>] [--spend|--view] <filename>

Command description:
  Sign the contents of a file with the given subaddress (or the main address if not specified)
```
where `--spend/--view` is used to choose which key to sign the transaction.

I updated the documentation to reflect that.

Reply:
I don't think view key can transfer assets although it can sign transactions.

Reply:
It is also my understanding.